### PR TITLE
Emails to s3 add alarms

### PR DESCRIPTION
**Pull request by david-pepper**
## What does this change?
Add alarms that fire under the following circumstances:

- failed to retrieve records from Salesforce (failure on getConfig, authentication with Salesforce, GET records)
- failed to check if s3 file exists
- failed to get file from s3
- failed to write file to s3
- failed to write back successes to Salesforce (high level - e.g. malformed json)
- failed to write back success to Salesforce (low level - e.g. malformed id)

An alarm should fire if 1 of these events occurs within a 1 hour period

## How to test
Each of the scenarios above were verified using the following framework: [Lambda Alarming - Test Plan](https://docs.google.com/spreadsheets/d/1oZ7AnTvP8WoMeKQo2JTO1OnJlOHyvmWhfPkYAfHpRMQ/edit?usp=sharing)

### CODE REVIEW

File: handlers/sf-emails-to-s3-exporter/cfn.yaml [hunk 1]
```diff
@@ -313,3 +318,142 @@ Resources:
               Action: s3:GetObject
               Resource:
                 - arn:aws:s3::*:membership-dist/*
+
+  failedRunAlarm:
+    Type: AWS::CloudWatch::Alarm
+    Condition: IsDev
```
Comment (kelvin-chappell) [inline handlers/sf-emails-to-s3-exporter/cfn.yaml:324]: Is this just for testing?

File: handlers/sf-emails-to-s3-exporter/cfn.yaml [hunk 2]
```diff
@@ -313,3 +318,142 @@ Resources:
...
+      Dimensions:
+        - Name: FunctionName
+          Value: !Ref LambdaFunction
+      EvaluationPeriods: 1
+      MetricName: failed_s3_check_file_exists
+      Namespace: s3-emails-from-sf
+      Period: 3600
+      Statistic: Sum
+      Threshold: 1
+      TreatMissingData: notBreaching
```
Comment (kelvin-chappell) [inline handlers/sf-emails-to-s3-exporter/cfn.yaml:366]: Might want to be `ignore` if you don't want the alarm state to change until a problem is resolved.

File: handlers/sf-emails-to-s3-exporter/cfn.yaml [hunk 3]
```diff
@@ -313,3 +318,142 @@ Resources:
...
+      Dimensions:
+        - Name: FunctionName
+          Value: !Ref LambdaFunction
+      EvaluationPeriods: 1
+      MetricName: Errors
+      Namespace: AWS/Lambda
+      Period: 3600
+      Statistic: Sum
+      Threshold: 1
+      TreatMissingData: notBreaching
```
Comment (kelvin-chappell) [inline handlers/sf-emails-to-s3-exporter/cfn.yaml:343]: Might want to be `ignore` if you don't want the alarm state to change until a problem is resolved.

File: handlers/sf-emails-to-s3-exporter/cfn.yaml [hunk 4]
```diff
@@ -313,3 +318,142 @@ Resources:
...
+      Dimensions:
+        - Name: FunctionName
+          Value: !Ref LambdaFunction
+      EvaluationPeriods: 1
+      MetricName: failed_s3_write_file
+      Namespace: s3-emails-from-sf
+      Period: 3600
+      Statistic: Sum
+      Threshold: 1
+      TreatMissingData: notBreaching
```
Comment (kelvin-chappell) [inline handlers/sf-emails-to-s3-exporter/cfn.yaml:412]: Might want to be `ignore` if you don't want the alarm state to change until a problem is resolved.

File: handlers/sf-emails-to-s3-exporter/cfn.yaml [hunk 5]
```diff
@@ -313,3 +318,142 @@ Resources:
...
+      Dimensions:
+        - Name: FunctionName
+          Value: !Ref LambdaFunction
+      EvaluationPeriods: 1
+      MetricName: failed_writeback_request_to_sf
+      Namespace: s3-emails-from-sf
+      Period: 3600
+      Statistic: Sum
+      Threshold: 1
+      TreatMissingData: notBreaching
```
Comment (kelvin-chappell) [inline handlers/sf-emails-to-s3-exporter/cfn.yaml:435]: Might want to be `ignore` if you don't want the alarm state to change until a problem is resolved.

File: handlers/sf-emails-to-s3-exporter/cfn.yaml (hunk 3 above)
Comment (david-pepper) [reply handlers/sf-emails-to-s3-exporter/cfn.yaml:344]: Done

File: handlers/sf-emails-to-s3-exporter/cfn.yaml (hunk 4 above)
Comment (david-pepper) [reply handlers/sf-emails-to-s3-exporter/cfn.yaml:416]: Done

File: handlers/sf-emails-to-s3-exporter/cfn.yaml (hunk 5 above)
Comment (david-pepper) [reply handlers/sf-emails-to-s3-exporter/cfn.yaml:440]: Done

File: handlers/sf-emails-to-s3-exporter/cfn.yaml [hunk 6]
```diff
@@ -313,3 +318,142 @@ Resources:
...
+      Dimensions:
+        - Name: FunctionName
+          Value: !Ref LambdaFunction
+      EvaluationPeriods: 1
+      MetricName: failed_writeback_to_sf_record
+      Namespace: s3-emails-from-sf
+      Period: 3600
+      Statistic: Sum
+      Threshold: 1
+      TreatMissingData: notBreaching
```
[the remark this answers is not in the document]
Comment (david-pepper) [reply handlers/sf-emails-to-s3-exporter/cfn.yaml]: Done

File: handlers/sf-emails-to-s3-exporter/src/main/scala/com/gu/sf_emails_to_s3_exporter/Metrics.scala [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+import com.gu.aws.AwsCloudWatch
+import com.gu.aws.AwsCloudWatch._
+
+import scala.util.Try
+
+object Metrics {
+
+  private val stage = sys.env.getOrElse("Stage", "DEV")
+
+  def put(event: String, value: Double = 1.0): Try[Unit] = {
```
[the remark this answers is not in the document]
Comment (david-pepper) [reply handlers/sf-emails-to-s3-exporter/src/main/scala/com/gu/sf_emails_to_s3_exporter/Metrics.scala]: Done

File: handlers/sf-emails-to-s3-exporter/cfn.yaml (hunk 1 above)
Comment (david-pepper) [reply handlers/sf-emails-to-s3-exporter/cfn.yaml]: yes, now testing has completed successfully I've updated the value to isProd

**Pull request merged**